### PR TITLE
Implement `wasp info` command

### LINKED ISSUE #277: Implement `wasp info` command
`wasp info` should tell you basic info about the project `wasp info` command was run from!

It should certainly write out the project name.

It could also provide some information like: What kind of database it uses, how big is the project, when was it compiled last time, ... .

We might also consider just `wasp` command always printing if it is inside a wasp project or not, and if so, what is the name of the project.

### PR DESCRIPTION
**Pull request by zant**
# Description

This PR implements `wasp info` command. It prints the following:

Screen Shot 2021-08-15 at 02 29 43

It also adds `info` and `deps` (`deps` was missing) to CLI Reference on the documentation.

**Information shown**
- Name: Taken from `main.wasp` using `getApp`.
- Compilation: Taken from `.wasp/out/.waspinfo` file. If it was not yet created prints "No compile information found".
- Project size: Using [dir-traverse](https://github.com/vmchale/dir-traverse) prints a sum of all file sizes. 

I'm super new to Haskell and Wasp, any suggestions will be highly appreciated. 

Wording of errors/info states can definitely be better. Also, the function to get all file sizes can probably be implemented without an external dependency. If the project prefers not introducing external libraries in minimal cases like this one, I'll be happy to reimplement it.

**Questions**
- What else can be printed?
- Since this command is related to project info, should it mention `wasp deps` for further details?

Thanks a lot!

Fixes #277

## Type of change

Please select the option(s) that is more relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### CODE REVIEW

File: waspc/cli/Command/Info.hs
Comment (Martinsos) [inline waspc/cli/Command/Info.hs:58]: I suggest we modify this function so it takes path to waspDir as an argument, and then it figures out on its own where the dotWaspFile is in it, reads it, and returns its contents. Why? Well, if its responsibility is to read a specific file (which it is), then it makes sense to move the piece of logic for finding that file inside of it, otherwise it is not really doing much.

Other thing is -> I would expect it to return content of the file, but sometimes it also returns "No compile information found". It means it returns two different things (content vs "error message") while pretending it is one thing, which is not great. Either we change the name of the function to be `readCompileInformation` and then it is better, or we keep the name as it is and instead return `IO (Maybe String)`, and then the external code will decide to print `No compile information found` when `Nothing` is returned.

This is all about names and responsibilities.

File: waspc/cli/Main.hs
Comment (Martinsos) [inline waspc/cli/Main.hs:89]:
```suggestion
        cmd "    info                  Prints basic information about current Wasp project.",
```

File: web/docs/cli.md
Comment (Martinsos) [inline web/docs/cli.md:26]: Update accordingly to previous comment.

File: waspc/stack.yaml
Comment (Martinsos) [inline waspc/stack.yaml:45]: Why is this here and not in package.yaml?

By the way, maybe you could use https://github.com/wasp-lang/wasp/blob/master/waspc/src/Util/IO.hs#L27 instead, so we don't have to add this extra dependency?

**Review by Martinsos — COMMENTED** [review]
Thanks for taking this on @zant, looks very good :)!!

I left some comments, nothing major, check them out.

Regarding the information printed: I like what you did, I think that is the right direction. Maybe we could rename `Compilation` to `Last compile` or `Last compilation`?
Regarding the size: next step could be showing the size of source code vs generated code (right now we show sum of those). Source code is everything without .wasp dir. Hm maybe we should skip all dirs starting with `.`? Because there could be `.git` and similar. And generated code - well that is everything inside `.wasp/out`. We can leave this for some future PR though.

What else could be interesting? Maybe some basic, fun information, like number of pages, number of entities, number of queries/operations, ... . But that is all also for the future PR, and it is hard to say how useful it is hm, although I think it would be fun.

All together, great work, especially since you don't have much experience with Haskell, I would say you did really well! Let's take care of that comments!

File: waspc/cli/Command/Info.hs
Comment (zant) [reply waspc/cli/Command/Info.hs]: Fair enough 👌

File: waspc/stack.yaml
Comment (zant) [reply waspc/stack.yaml:45]: This is exactly what I needed! Thanks you, will update properly.

File: waspc/cli/Command/Info.hs
Comment (zant) [reply waspc/cli/Command/Info.hs]: This is great review, thank you.

Regarding the first point, I completely agree. I believe I first moved it to a function just to add more readability on the main one. But moving the logic makes complete sense.

Regarding the second one, this is also super nice. I honestly wasn't taking that into consideration. At the end I decided to go with changing name to `readCompileInformation` as I'm explicitly printing the result as information - that it wasn't found.

**Comment by zant** [comment]
Hey @Martinsos this is amazing! Thanks a lot for the review and comments. I just updated the PR with the new changes and it's now ready for second review.

I really liked your suggestions about what and how to display the information. It can be lots of fun indeed, and count me in if a follow-up is needed to work on it!

And thanks again, I've learnt tons working on this PR!

File: waspc/cli/Command/Info.hs
Comment (Martinsos) [inline waspc/cli/Command/Info.hs:55]: use P.fromRelFile instead of P.toFilePath, as it gives extra type safety by ensuring we are dealing with files.

File: waspc/cli/Command/Info.hs
Comment (Martinsos) [inline waspc/cli/Command/Info.hs:59]:
```suggestion
  let dotWaspInfoFile = fromAbsFile $ waspDir </> Cli.Common.dotWaspDirInWaspProjectDir </> Cli.Common.generatedCodeDirInDotWaspDir </> Cli.Common.dotWaspInfoFileInGeneratedCodeDir
```

Again, more type safety.

File: waspc/cli/Command/Info.hs
Comment (Martinsos) [inline waspc/cli/Command/Info.hs:62]: One tricky thing here is that, if we modify waspinfo file in the future to contain more information than just compile info, we will probably forget about this piece of code here and will still be printing all of it here and it will probably be a bunch of non-revelant information. But ok, I don't see a simple way to improve that right now. We could introduce some kind of abstraction over waspinfo file that implements a function that returns only compile info from its contents (for now that would be `id`), and then it will be more obvious when we change it that we have that assumption, but I wouldn't waste time on it at the moment.

File: waspc/cli/Command/Info.hs
Comment (Martinsos) [inline waspc/cli/Command/Info.hs:71]:
```suggestion
      waspStr <- readFile (fromAbsFile waspFile)
```

File: waspc/cli/Command/Info.hs
Comment (Martinsos) [inline waspc/cli/Command/Info.hs:74]:
```suggestion
      return $ left (("Couldn't parse .wasp file: " <>) . show) $ Parser.parseWasp waspStr
```

I believe you can do this here, to make this shorter! `left` is coming from `Control.Arrow`.

**Review by Martinsos — COMMENTED** [review]
Awesome, glad to hear you liked the review :)!
I like the changes, I think we are there. I read your replies and it all makes sense.
I left a couple more comments but smaller this time -> let's take care of those and I will merge the PR!

File: waspc/cli/Command/Info.hs
Comment (zant) [reply waspc/cli/Command/Info.hs]: Cool one :)

File: waspc/cli/Command/Info.hs
Comment (zant) [reply waspc/cli/Command/Info.hs]: Oh yes! I remember having a runtime error related that I believe could be related to this.

File: waspc/cli/Command/Info.hs
Comment (zant) [reply waspc/cli/Command/Info.hs:63]: Oh yes I see, it makes total sense. Maybe expanding `.waspinfo` to some markup language to parse easily, like a yaml file with some keys. Or an easy format we can parse, assuming that `.waspinfo` will grow and some other info will be there and granular access will be needed.

The other solution that doesn't involves introducing new markup or formats of some sort can be using the modification date of `.waspinfo`. Making the assumption that, the last time that this file was modified, it was also the last time that it was compiled. 

However this can also mean that manual modification will also change compilation information, which is not optimal.

But yes, as you mentioned, I believe that an abstraction to not forget it, will be reasonable way to go at the moment. 👍

**Comment by zant** [comment]
Last changes updated and ready for review! 

Good comments again. I'm now more motivated to keep on reading about Haskell concepts and working on more projects with it. 

I'm also excited to try Wasp at the app level. I think being able to leverage different backends can be really cool. For example I was thinking about a Web3 backend interfacing with smart contracts, IMO can simply things a lot.

Really cool stuff!  👌

File: waspc/cli/Command/Info.hs
Comment (Martinsos) [reply waspc/cli/Command/Info.hs:63]: That is a way to do it also! But as you said, we are making assumptions nobody else is touching it + that dotwaspinfo will be changed only on compile, which might not be true in the future, wasp might be storing additional information there.

I don't think we should be solving this now, just wanted to give heads up.

**Comment by Martinsos** [comment]
> Last changes updated and ready for review!
> 
> Good comments again. I'm now more motivated to keep on reading about Haskell concepts and working on more projects with it 
> 
> Oh and I'm also excited to try Wasp at the app level. I think being able to leverage different backends can be really cool. For example I was thinking about a Web3 backend interfacing with smart contracts to expose its methods. IMO it can simplify things a lot on that end.
> 
> Really cool stuff!

Ha, well I am also excited that you are excited, both about Haskell and Wasp :D! Great to have you contribute and I hope you will want to contribute some more in the future :).

Although we will probably not be getting into it before we release a stable version of React + Node support, I agree that multiple backends could be very cool! I think it is hard to guess how that will play out and to what extend will people accept them and how exactly will they use them, but I think it is an exciting direction to explore.

So this PR is ready to be merged but, as you can see above, the code formatter check is failing. That is because we are using a specific formater, [ormolu](https://github.com/tweag/ormolu), so you will also have to use it to format this piece of code. Easiest way is to install an extension for your editor that will install/use it, normally it triggers formatting on file save. Or, you could install ormolu and use it from CLI to format the code, that is also an option. But I would suggest going with the first option, so you have it also for the rest of your Haskell projects. Once you fix that and checks pass, I will merge!

**Comment by zant** [comment]
That's great @Martinsos. Thanks for elaborating on that! 

Just installed ormolu + VSCode and updated the PR accordingly. Hopefully it will now pass all checks 🙌

And for surely will like to keep contributing, looking forward already! Haha

**Review by Martinsos — APPROVED** [review]

**Comment by Martinsos** [comment]
Awesome, that is it! Merged! Btw wow 30 commits, that is impressive granulation :D. I squashed them into one big commit and rebased on master.

**Comment by zant** [comment]
Yayyyy! Hahah yes I'm a bit of a commiter. That's great! Super happy, thanks a lot @Martinsos :D

**Pull request merged**
